assets/js/googlesitekit/modules/datastore/modules.js

```diff
@@ -307,9 +379,60 @@ const baseResolvers = {
 
 		if ( ! existingModules ) {
 			yield fetchGetModulesStore.actions.fetchGetModules();
+			yield registry.dispatch( STORE_NAME ).invalidateResolutionForStoreSelector( 'canActivateModule' );
+			yield registry.dispatch( STORE_NAME ).invalidateResolutionForStoreSelector( 'getCheckRequirementsStatus' );
```

Comment:
Following up on the previous comments, these invalidations shouldn't happen here for `getModules`. They need to happen when a new module is registered in JS, which is the action that _may_ change the outcome of these selectors by registering the `checkRequirements` function.

In other words, we need to move this to `registerModule`. In `registerModule`, we can make it more specific too, to avoid invalidating "irrelevant" resolvers: We only need to invalidate resolutions for the selector for the specifically registered module, i.e. `yield registry.dispatch( STORE_NAME ).invalidateResolution( 'canActivateModule', [ slug ] );` and `yield registry.dispatch( STORE_NAME ).invalidateResolution( 'getCheckRequirementsStatus', [ slug ] );`.

Reply:
Are we sure that we want to do this in registerModule action? Since the module is only registered (for the purpose of getModule() selector) after the reducer REGISTER_MODULE completes, it's only after that point that we should be invalidating. In the registerModule would be too early.

Reply:
Good point, we cannot do it the way that action is currently implemented. However, the call needs to happen in `registerModule` because that is where a `checkRequirements` function may be registered which would alter the results. We should be able to implement that though without too much effort:
* Make `registerModule` a generator.
* `yield` the `REGISTER_MODULE` action data.
* This way we can _afterwards_ in the `registerModule` action invalidate the resolutions for these selectors and the given `slug`.

You can check the `setModuleActivation` implementation in the same file for a fairly similar approach (the only difference is that `REGISTER_MODULE` doesn't return anything).

Reply:
Unfortunately it's still too soon, I've changed it to a generator, added the resolver invalidation and added some quick test code to `getCheckRequirementsError` for the slug, with the expectation that `optimize` will give us this error:
```json
{
    "code": "insufficient_module_dependencies", 
    "message": "You need to set up Analytics to gain access to Optimize.", 
    "data": {
        "inactiveModules": [
            "Analytics"
        ]
    }
}
```

```js

		yield {
			payload: {
				settings,
				slug,
			},
			type: REGISTER_MODULE,
		};

		const registry = yield Data.commonActions.getRegistry();

		yield registry.dispatch( STORE_NAME ).invalidateResolution( 'canActivateModule', [ slug ] );
		yield registry.dispatch( STORE_NAME ).invalidateResolution( 'getCheckRequirementsError', [ slug ] );
		console.log( slug ); // eslint-disable-line no-console
		// eslint-disable-next-line no-undef
		console.log( googlesitekit.data.stores[ 'core/modules' ].getSelectors().getCheckRequirementsError( slug ) ); // eslint-disable-line no-console
```

However all of the errors are still undefined.

Reply:
On second thought, the likelihood of running one of those resolvers immediately is very small (or zero?) so I pushed up the generator change in https://github.com/google/site-kit-wp/pull/2335/commits/206ffa353a511394bcbcfb49016f73d4000a8ace